### PR TITLE
Fix locking of sdists rejected by Pip.

### LINKED ISSUE #2519: `pex3 lock sync` isn't a no-op when it seems like it should be
This is on MacOS on an M1:

```
$ rm -f lockfile.json
$ python -m pex.cli lock sync --style strict --interpreter-constraint $'CPython>=3.10,<3.11' --indent 2 --lock lockfile.json  'bloom-filter2>=2.0.0'
$ python -m pex.cli lock sync --style strict --interpreter-constraint $'CPython>=3.10,<3.11' --indent 2 --lock lockfile.json  'bloom-filter2>=2.0.0'
Updates for lock generated by cp310-cp310-macosx_14_0_arm64:
  Updated bloom-filter2 2 artifacts:
    + https://files.pythonhosted.org/packages/52/d7/871408c1c1230af64eb00e2064f6e8a065ae652f17f010b3f4f102e5fbac/bloom_filter2-2.0.0-py3-none-any.whl
    - https://files.pythonhosted.org/packages/2b/5d/8de4a849ebe212217e6d8f4798a6918d4035e741c44730da81272f170b47/bloom-filter2-2.0.0-1.tar.gz
$ python -m pex.cli lock sync --style strict --interpreter-constraint $'CPython>=3.10,<3.11' --indent 2 --lock lockfile.json  'bloom-filter2>=2.0.0'
No updates for lock generated by cp310-cp310-macosx_14_0_arm64.
$ python -m pex.cli lock sync --style strict --interpreter-constraint $'CPython>=3.10,<3.11' --indent 2 --lock lockfile.json  'bloom-filter2>=2.0.0'
No updates for lock generated by cp310-cp310-macosx_14_0_arm64.
$ 
```

The first invocation generates the lockfile. I would expect the second to be a no-op, since no inputs have changed (and neither has the state on PyPI). Yet it updates the artifact, as logged. The diff between the two lockfile versions confirms this:

```
15,16c15,16
<               "hash": "310e5d3aa15be6e3d2285a7111f60cc44a1f500159412f2c72d155f43037cefd",
<               "url": "https://files.pythonhosted.org/packages/2b/5d/8de4a849ebe212217e6d8f4798a6918d4035e741c44730da81272f170b47/bloom-filter2-2.0.0-1.tar.gz"
---
>               "hash": "f9d3a3a44ae917ab9c13f71b542ac88b863f0cf4e64afbca48777cf0d5310c93",
>               "url": "https://files.pythonhosted.org/packages/52/d7/871408c1c1230af64eb00e2064f6e8a065ae652f17f010b3f4f102e5fbac/bloom_filter2-2.0.0-py3-none-any.whl"
```

I'll note that this is one of possibly a couple of odd phenomena I noticed while testing `lock sync` on a large set of input requirements, but this was the easiest one to isolate in a small example. I will look further into this and the others, but wanted to file this to record my findings.

### PR DESCRIPTION
**Pull request by jsirois**
Previously, the Pex locking logic would miss when an sdist was initially
picked by Pip, but then rejected and a wheel was chosen instead.

Fixes #2519

### CODE REVIEW

File: tests/integration/cli/commands/test_issue_2519.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+@skip_if_py2
+@pytest.mark.parametrize(
+    "pip_version",
+    [
+        pytest.param(version, id=str(version))
+        for version in PipVersion.values()
+        if version >= PipVersion.v23_2 and version.requires_python_applies(TARGET)
+    ],
+)
+def test_sync_version_mismatch_new_pip(
```
Comment (jsirois) [inline tests/integration/cli/commands/test_issue_2519.py:107]: This tests the case that was previously failing with:
```console
:; python -m pex.cli lock sync --style strict --interpreter-constraint "CPython>=3.10,<3.11" --indent 2 --lock lockfile.json  "bloom-filter2>=2.0.0" --pip-version 23.2 -v
pex: Hashing pex: 18.5ms
pex: Isolating pex: 0.1ms
pex: Resolving for:
pex: Resolving for:inux_2_39_x86_64 interpreter at /home/jsirois/.pyenv/versions/3.10.14/bin/python3.10
  cp310-cp310-manylinux_2_39_x86_64 interpreter at /home/jsirois/.pyenv/versions/3.10.14/bin/python3.10: 923.7ms
pex: Creating lock from resolve: 13.0ms
pex:   Building 0 source distributions to gather metadata for lock.: 8.6ms
pex: Indexing downloads: 0.3ms
Traceback (most recent call last):
  File "/home/jsirois/dev/pex-tool/pex/pex/result.py", line 105, in catch
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/pex-tool/pex/pex/cli/command.py", line 110, in run
    return subcommand_func(self)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/pex-tool/pex/pex/cli/commands/lock.py", line 1567, in _sync
    create(
  File "/home/jsirois/dev/pex-tool/pex/pex/resolve/lockfile/create.py", line 438, in create
    create_lock_download_manager = CreateLockDownloadManager.create(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jsirois/dev/pex-tool/pex/pex/resolve/lockfile/create.py", line 100, in create
    artifact_and_project_name = source_artifacts_by_pin[pin]
                                ~~~~~~~~~~~~~~~~~~~~~~~^^^^^
KeyError: Pin(project_name=ProjectName(raw='bloom_filter2', validated=False, normalized='bloom-filter2'), version=Version(raw='2.0.0', normalized='2'))
```

File: tests/integration/cli/commands/test_issue_2519.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+@skip_if_py2
+@pytest.mark.parametrize(
+    "pip_version",
+    [
+        pytest.param(version, id=str(version))
+        for version in PipVersion.values()
+        if version < PipVersion.v23_2 and version.requires_python_applies(TARGET)
+    ],
+)
+def test_sync_version_mismatch_old_pip(
```
Comment (jsirois) [inline tests/integration/cli/commands/test_issue_2519.py:61]: This just asserts the OP case explanation offered in https://github.com/pex-tool/pex/issues/2519#issuecomment-2341715425.

File: tests/integration/cli/commands/test_issue_2519.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+@skip_if_py2
+@pytest.mark.parametrize(
+    "pip_version",
+    [
+        pytest.param(version, id=str(version))
+        for version in PipVersion.values()
+        if version.requires_python_applies(TARGET)
+    ],
+)
+def test_sync_version_mismatch_prefer_binary_all_pips(
```
Comment (jsirois) [inline tests/integration/cli/commands/test_issue_2519.py:138]: This test demonstrates the alternate workaround to using a newer Pip described in #2519, using `--prefer-binary` instead. Note that this works over all Pip versions, but is not typically generic. You'd only want to apply it for certain resolves.

File: pex/resolve/lockfile/updater.py [hunk 4]
```diff
@@ -619,6 +619,7 @@ def create(
         max_jobs,  # type: int
         use_pip_config,  # type: bool
         dependency_configuration,  # type: DependencyConfiguration
+        preserve_log,  # type: bool
```
Comment (jsirois) [inline pex/resolve/lockfile/updater.py:622]: When investigating this I needed to examine Pip logs. The `--preserve-pip-download-log` wasn't working for `pex3 lock sync` run 2+ which prompted this otherwise unrelated fix in this file and the `pex/resolve/resolver_options.py` below.

**Review by benjyw — APPROVED** [review]
